uptane/common.py

```diff
@@ -62,21 +90,136 @@ def sign_signable(signable, keys_to_sign_with):
 
     # Else, all is well. Sign the signable with the given key, adding that
     # signature to the signatures list in the signable.
-    signable['signatures'].append(
-        tuf.keys.create_signature(
-        signing_key, signable['signed'], force_treat_as_pydict=True))
+    signable['signatures'].append(sign_over_metadata(
+        signing_key, signable['signed'], datatype=datatype,
+        metadata_format=metadata_format))
+
+
+  uptane.formats.ANY_SIGNABLE_UPTANE_METADATA_SCHEMA.check_match(signable)
+
+  return
+
+
+
+
+
+def sign_over_metadata(
+    key_dict, data, datatype, metadata_format=tuf.conf.METADATA_FORMAT):
+  """
+  Almost exactly identical to the function simultaneously added to TUF,
+  tuf.sig.sign_over_metadata(). Requires datatype.
+  Must differ in Uptane simply because it is not possible to convert
+  Uptane-specific metadata (Time Attestations, ECU Manifests, and Vehicle
+  Manifests) to or from ASN.1/DER without knowing which of those three
+  types of metadata you're dealign with, and this conversion is required for
+  signing and verifying signatures.
+
+  Higher level function that wraps tuf.keys.create_signature, and works
+  specifically with Time Attestations, ECU Manifsts, and Vehicle Manifests that
+  will be in JSON or ASN.1/DER format.
+
+  See tuf.keys.create_signature for overall functionality and the arguments
+  key_dict and data.
+
+  Optional argument:
+
+    metadata_format: (optional; default based on tuf.conf.METADATA_FORMAT)
+
+      If 'json', treats data as a JSON-friendly Python dictionary to be turned
+      into a canonical JSON string and then encoded as utf-8 before signing.
+      When operating TUF with DER metadata but checking the signature on some
+      piece of JSON for some reason, this should be manually set to 'json'. The
+      purpose of this canonicalization is to produce repeatable signatures
+      across different platforms and Python key dictionaries (avoiding things
+      like different signatures over the same dictionary).
+
+      If 'der', the data will be converted into ASN.1, encoded as DER,
+      and hashed. The signature is then checked against that hash.
+  """
+
+  tuf.formats.ANYKEY_SCHEMA.check_match(key_dict)
+  # TODO: Check format of data, based on metadata_format.
+  # TODO: Consider checking metadata_format redundantly. It's checked below.
+
+  if metadata_format == 'json':
+    data = tuf.formats.encode_canonical(data).encode('utf-8')
+
+  elif metadata_format == 'der':
+
+    # TODO: Have convert_signed_metadata_to_der take just the 'signed' element
+    # so we don't have to do this silly wrapping in an empty signable.
+    data = asn1_codec.convert_signed_metadata_to_der(
+        {'signed': data, 'signatures': []}, only_signed=True, datatype=datatype)
+    data = hashlib.sha256(data).digest()
```

Comment:
Agreed. This particular hashing I see as a temporary measure, though: it remains here because some lower-level signing modules in TUF are signing hashes, and others aren't. (Incidentally, sha256 is hard-coded in some of those, too.) The other original motivation for this (avoiding some encoding issues) his no longer a concern.

Wherever the hashing ends up happening, it should be configurable.